zebra-consensus/src/checkpoint.rs

```diff
@@ -3,315 +3,1217 @@
+            FinalCheckpoint => None,
+        }
+    }
+
+    /// Return the target checkpoint height that we want to verify.
+    ///
+    /// If we need more blocks, returns `WaitingForBlocks`.
+    ///
+    /// If the queued blocks are continuous from the previous checkpoint to a
+    /// subsequent checkpoint, returns `Checkpoint(height)`. The target
+    /// checkpoint can be multiple checkpoints ahead of the previous checkpoint.
+    /// `height` increases as checkpoints are verified.
+    ///
+    /// If verification has finished, returns `FinishedVerifying`.
+    fn target_checkpoint_height(&self) -> Target<BlockHeight> {
+        // Find the height we want to start searching at
+        let mut pending_height = match self.previous_checkpoint_height() {
+            // Check if we have the genesis block as a special case, to simplify the loop
+            BeforeGenesis if !self.queued.contains_key(&BlockHeight(0)) => return WaitingForBlocks,
+            BeforeGenesis => BlockHeight(0),
+            PreviousCheckpoint(height) => height,
+            FinalCheckpoint => return FinishedVerifying,
+        };
+
+        // Find the end of the continuous sequence of blocks, starting at the
+        // last verified checkpoint. If there is no verified checkpoint, start
+        // *after* the genesis block (which we checked above).
+        //
+        // If `btree_map::Range` implements `ExactSizeIterator`, it would be
+        // much faster to walk the checkpoint list, and compare the length of
+        // the `btree_map::Range` to the block height difference between
+        // checkpoints. (In maps, keys are unique, so we don't need to check
+        // each height value.)
+        //
+        // But at the moment, this implementation is slightly faster, because
+        // it stops after the first gap.
+        for (&height, _) in self.queued.range((Excluded(pending_height), Unbounded)) {
+            // If the queued blocks are continuous.
+            //
+            // Precondition: queued block heights are within the valid
+            // BlockHeight range, so there is no integer overflow.
+            assert!(
+                pending_height.0 < u32::MAX,
+                "pending block had a height of u32::MAX "
+            );
+            if height == BlockHeight(pending_height.0 + 1) {
+                pending_height = height;
+            } else {
+                break;
+            }
+        }
+
+        // Now find the start of the checkpoint range
+        let start = self.current_start_bound().expect(
+            "if verification has finished, we should have returned earlier in the function",
+        );
+        // Find the highest checkpoint below pending_height, excluding any
+        // previously verified checkpoints
+        let target_checkpoint = self
+            .checkpoint_list
+            .max_height_in_range((start, Included(pending_height)));
+
+        if let Some(target_checkpoint) = target_checkpoint {
```

Comment:
TIOLI: I generally prefer to use combinators on option in cases like this

```rust
target_checkpoint
    .map(Checkpoint)
    .unwrap_or(WaitingForBlocks)
```

Reply:
Fixed in 3540e713804e870c1ec0b7855ebd518a213d532f